pkg/handler/topology.go

```diff
@@ -47,7 +47,27 @@ func (h *Handlers) GetTopology(ctx context.Context) func(w http.ResponseWriter,
 			metrics.ObserveHTTPCall("GetTopology", code, startTime)
 		}()
 
-		flows, code, err := h.getTopologyFlows(ctx, clients, r.URL.Query())
+		params := r.URL.Query()
+		ds, err := getDatasource(params)
+		if err != nil {
+			writeError(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		flows, code, err := h.getTopologyFlows(ctx, clients, params, ds)
+		var dsErr *datasourceError
+		if err != nil &&
+			ds == constants.DataSourceAuto &&
+			h.Cfg.IsLokiEnabled() &&
+			(code == http.StatusForbidden || code == http.StatusUnauthorized) &&
+			errors.As(err, &dsErr) &&
+			dsErr.datasource == constants.DataSourceProm {
+			// In case this was a prometheus 401 / 403 error, the query is repeated with Loki
+			// This is because multi-tenancy is currently not managed for prom datasource, hence such queries have to go with Loki
+			// Unfortunately we don't know a safe and generic way to pre-flight check if the user will be authorized
+			hlog.Info("Retrying with Loki...")
+			flows, code, err = h.getTopologyFlows(ctx, clients, params, constants.DataSourceLoki)
+		}
```

Comment:
why not using something like `wait.PollUntilContextTimeout` to implement proper retry instead of retrying just once ?

Reply:
It's not a "Retry until something", but replaying same query while switching to the other datasource